File: app/src/main/java/com/mdgd/pokemon/ui/pokemon/PokemonDetailsViewModel.java
package com.mdgd.pokemon.ui.pokemon;

import androidx.lifecycle.Lifecycle;

import com.mdgd.mvi.MviViewModel;
import com.mdgd.pokemon.R;
import com.mdgd.pokemon.models.repo.PokemonsRepo;
import com.mdgd.pokemon.models.repo.dao.schemas.PokemonFullDataSchema;
import com.mdgd.pokemon.models.repo.dao.schemas.PokemonSchema;
import com.mdgd.pokemon.models.repo.schemas.Ability;
import com.mdgd.pokemon.models.repo.schemas.Form;
import com.mdgd.pokemon.models.repo.schemas.GameIndex;
import com.mdgd.pokemon.models.repo.schemas.Stat;
import com.mdgd.pokemon.models.repo.schemas.Type;
import com.mdgd.pokemon.ui.pokemon.items.ImagePropertyData;
import com.mdgd.pokemon.ui.pokemon.items.LabelPropertyData;
import com.mdgd.pokemon.ui.pokemon.items.PokemonProperty;
import com.mdgd.pokemon.ui.pokemon.items.TextPropertyData;
import com.mdgd.pokemon.ui.pokemon.items.TitlePropertyData;
import com.mdgd.pokemon.ui.pokemon.state.PokemonScreenState;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.schedulers.Schedulers;
import io.reactivex.rxjava3.subjects.BehaviorSubject;

public class PokemonDetailsViewModel extends MviViewModel<PokemonDetailsContract.View, PokemonScreenState> implements PokemonDetailsContract.ViewModel {

    private final BehaviorSubject<Long> pokemonIdSubject = BehaviorSubject.create();
    private final PokemonsRepo repo;

    public PokemonDetailsViewModel(PokemonsRepo repo) {
        this.repo = repo;
    }

    @Override
    public void setPokemonId(long pokemonId) {
        pokemonIdSubject.onNext(pokemonId);
    }

    @Override
    public void onStateChanged(Lifecycle.Event event) {
        super.onStateChanged(event);
        if (event == Lifecycle.Event.ON_CREATE && !hasOnDestroyDisposables()) {
            observeTillDestroy(pokemonIdSubject
                    .switchMap(repo::getPokemonsById)
                    .map(optional -> optional.isPresent() ? mapToListPokemon(optional.get()) : new LinkedList<PokemonProperty>())
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(list -> setState(new PokemonScreenState.SetItems(list))));
        }
    }

    private List<PokemonProperty> mapToListPokemon(PokemonFullDataSchema pokemonDetails) {
        final List<PokemonProperty> properties = new ArrayList<>();
        final PokemonSchema pokemonSchema = pokemonDetails.getPokemonSchema();
        properties.add(new ImagePropertyData(pokemonSchema.getSprites().getOther().getOfficialArtwork().getFrontDefault()));
        properties.add(new LabelPropertyData(R.string.pokemon_detail_name, pokemonSchema.getName()));
        properties.add(new LabelPropertyData(R.string.pokemon_detail_height, String.valueOf(pokemonSchema.getHeight())));
        properties.add(new LabelPropertyData(R.string.pokemon_detail_weight, String.valueOf(pokemonSchema.getWeight())));

        properties.add(new TitlePropertyData(R.string.pokemon_detail_stats));
        for (Stat s : pokemonDetails.getStats()) {
            properties.add(new LabelPropertyData(s.getStat().getName(), String.valueOf(s.getBaseStat()), 1));
        }
        properties.add(new TitlePropertyData(R.string.pokemon_detail_abilities));
        final List<Ability> abilities = pokemonDetails.getAbilities();
        final StringBuilder abilitiesText = new StringBuilder();
        for (int i = 0; i < abilities.size(); i++) {
            abilitiesText.append(abilities.get(i).getAbility().getName());
            if (i < abilities.size() - 1) {
                abilitiesText.append(", ");
            }
        }
        properties.add(new TextPropertyData(abilitiesText.toString(), 1));

        properties.add(new TitlePropertyData(R.string.pokemon_detail_forms));
        final List<Form> forms = pokemonDetails.getForms();
        final StringBuilder formsText = new StringBuilder();
        for (int i = 0; i < forms.size(); i++) {
            formsText.append(forms.get(i).getName());
            if (i < forms.size() - 1) {
                formsText.append(", ");
            }
        }
        properties.add(new TextPropertyData(formsText.toString(), 1));


        properties.add(new TitlePropertyData(R.string.pokemon_detail_types));
        final List<Type> types = pokemonDetails.getTypes();
        final StringBuilder typesText = new StringBuilder();
        for (int i = 0; i < types.size(); i++) {
            typesText.append(types.get(i).getType().getName());
            if (i < types.size() - 1) {
                typesText.append(", ");
            }
        }
        properties.add(new TextPropertyData(typesText.toString(), 1));


        properties.add(new TitlePropertyData(R.string.pokemon_detail_game_indicies));
        final List<GameIndex> gameIndices = pokemonDetails.getGameIndices();
        final StringBuilder gameIndicesText = new StringBuilder();
        for (int i = 0; i < gameIndices.size(); i++) {
            gameIndicesText.append(gameIndices.get(i).getVersion().getName());
            if (i < gameIndices.size() - 1) {
                gameIndicesText.append(", ");
            }
        }
        properties.add(new TextPropertyData(gameIndicesText.toString(), 1));
        return properties;
    }
}

File: app/src/main/java/com/mdgd/pokemon/ui/pokemon/PokemonDetailsViewModel.kt
package com.mdgd.pokemon.ui.pokemon

import androidx.lifecycle.Lifecycle
import androidx.lifecycle.viewModelScope
import com.mdgd.mvi.MviViewModel
import com.mdgd.pokemon.R
import com.mdgd.pokemon.models.repo.PokemonsRepo
import com.mdgd.pokemon.models.repo.dao.schemas.PokemonFullDataSchema
import com.mdgd.pokemon.models.repo.schemas.Ability
import com.mdgd.pokemon.models.repo.schemas.Form
import com.mdgd.pokemon.models.repo.schemas.GameIndex
import com.mdgd.pokemon.models.repo.schemas.Type
import com.mdgd.pokemon.ui.pokemon.dto.ImageProperty
import com.mdgd.pokemon.ui.pokemon.dto.LabelProperty
import com.mdgd.pokemon.ui.pokemon.dto.PokemonProperty
import com.mdgd.pokemon.ui.pokemon.dto.TextProperty
import com.mdgd.pokemon.ui.pokemon.dto.TitleProperty
import com.mdgd.pokemon.ui.pokemon.state.PokemonDetailsScreenState
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch
import java.util.LinkedList

class PokemonDetailsViewModel(
    private val repo: PokemonsRepo
) : MviViewModel<PokemonDetailsContract.View, PokemonDetailsScreenState>(),
    PokemonDetailsContract.ViewModel {

    private val pokemonIdFlow = MutableStateFlow(-1L)
    private var pokemonLoadingJob: Job? = null

    override fun setPokemonId(pokemonId: Long) {
        viewModelScope.launch {
            pokemonIdFlow.emit(pokemonId)
        }
    }

    override fun onStateChanged(event: Lifecycle.Event) {
        super.onStateChanged(event)

        if (event == Lifecycle.Event.ON_CREATE && pokemonLoadingJob == null) {
            pokemonLoadingJob = viewModelScope.launch {
                pokemonIdFlow
                    .filter { it != -1L }
                    .map { id: Long -> repo.getPokemonById(id) }
                    .map { details: PokemonFullDataSchema? ->
                        if (details == null) LinkedList() else mapToListPokemon(
                            details
                        )
                    }
                    .flowOn(Dispatchers.IO)
                    .collect { setState(PokemonDetailsScreenState.SetData(it)) }
            }
        }
    }

    private fun mapToListPokemon(pokemonDetails: PokemonFullDataSchema): List<PokemonProperty> {
        val properties: MutableList<PokemonProperty> = ArrayList()
        val pokemonSchema = pokemonDetails.pokemonSchema
        properties.add(ImageProperty(pokemonSchema!!.sprites!!.other!!.officialArtwork!!.frontDefault!!))
        properties.add(LabelProperty(R.string.pokemon_detail_name, pokemonSchema.name!!))
        properties.add(
            LabelProperty(
                R.string.pokemon_detail_height,
                pokemonSchema.height.toString()
            )
        )
        properties.add(
            LabelProperty(
                R.string.pokemon_detail_weight,
                pokemonSchema.weight.toString()
            )
        )
        properties.add(TitleProperty(R.string.pokemon_detail_stats))
        for (s in pokemonDetails.stats) {
            properties.add(LabelProperty(s.stat!!.name!!, s.baseStat.toString(), 1))
        }
        properties.add(TitleProperty(R.string.pokemon_detail_abilities))
        val abilities: List<Ability> = pokemonDetails.abilities
        val abilitiesText = StringBuilder()
        for (i in abilities.indices) {
            abilitiesText.append(abilities[i].ability!!.name)
            if (i < abilities.size - 1) {
                abilitiesText.append(", ")
            }
        }
        properties.add(TextProperty(abilitiesText.toString(), 1))
        properties.add(TitleProperty(R.string.pokemon_detail_forms))
        val forms: List<Form> = pokemonDetails.forms
        val formsText = StringBuilder()
        for (i in forms.indices) {
            formsText.append(forms[i].name)
            if (i < forms.size - 1) {
                formsText.append(", ")
            }
        }
        properties.add(TextProperty(formsText.toString(), 1))
        properties.add(TitleProperty(R.string.pokemon_detail_types))
        val types: List<Type> = pokemonDetails.types
        val typesText = StringBuilder()
        for (i in types.indices) {
            typesText.append(types[i].type!!.name)
            if (i < types.size - 1) {
                typesText.append(", ")
            }
        }
        properties.add(TextProperty(typesText.toString(), 1))
        properties.add(TitleProperty(R.string.pokemon_detail_game_indicies))
        val gameIndices: List<GameIndex> = pokemonDetails.gameIndices
        val gameIndicesText = StringBuilder()
        for (i in gameIndices.indices) {
            gameIndicesText.append(gameIndices[i].version!!.name)
            if (i < gameIndices.size - 1) {
                gameIndicesText.append(", ")
            }
        }
        properties.add(TextProperty(gameIndicesText.toString(), 1))
        return properties
    }

    override fun onCleared() {
        super.onCleared()
        pokemonLoadingJob = null
    }
}
